oval/redhat/redhat.go

```diff
@@ -0,0 +1,110 @@
+package redhat
+
+import (
+	"bytes"
+	"compress/bzip2"
+	"encoding/xml"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/afero"
+	"golang.org/x/xerrors"
+	"gopkg.in/cheggaaa/pb.v1"
+
+	"github.com/aquasecurity/vuln-list-update/utils"
+)
+
+const (
+	ovalDir   = "oval"
+	redhatDir = "redhat"
+	urlFormat = "https://www.redhat.com/security/data/oval/v2/RHEL%s/rhel-%s.oval.xml.bz2"
+	retry     = 5
+)
+
+var (
+	releases = []string{"6", "7", "8"}
+)
+
+type Config struct {
+	VulnListDir string
+	URLFormat   string
+	AppFs       afero.Fs
+	Retry       int
+}
+
+func NewConfig() Config {
+	return Config{
+		VulnListDir: utils.VulnListDir(),
+		URLFormat:   urlFormat,
+		AppFs:       afero.NewOsFs(),
+		Retry:       retry,
+	}
+}
+
+func (c Config) Update() error {
+	log.Println("Fetching Red Hat OVAL data...")
+	for _, release := range releases {
+		if err := c.update(release); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c Config) update(release string) error {
+	url := fmt.Sprintf(c.URLFormat, release, release)
+	res, err := utils.FetchURL(url, "", c.Retry)
+	if err != nil {
+		return xerrors.Errorf("failed to fetch Red Hat OVAL: %w", err)
+	}
+
+	bzr := bzip2.NewReader(bytes.NewBuffer(res))
+
+	log.Printf("Updating Red Hat %s OVAL data...\n", release)
+	ovalroot := Root{}
+	if err = xml.NewDecoder(bzr).Decode(&ovalroot); err != nil {
+		return xerrors.Errorf("failed to unmarshal Red Hat OVAL XML: %w", err)
+	}
+
+	dir := filepath.Join(ovalDir, redhatDir, release)
+	bar := pb.StartNew(len(ovalroot.Definitions.Definitions))
+	for _, def := range ovalroot.Definitions.Definitions {
+		titles := strings.Fields(def.Title)
+		title := strings.Trim(titles[0], ":")
+		if err = c.saveRHSAPerYear(dir, title, def); err != nil {
+			return xerrors.Errorf("unable to save RHEL advisory: %w", err)
+		}
+		bar.Increment()
+	}
+	bar.Finish()
+	return nil
+}
+
+func (c Config) saveRHSAPerYear(dirName string, rhsaID string, data interface{}) error {
+	// e.g. RHSA-2018:0094
+	s := strings.Split(rhsaID, ":")
```

Comment:
This might be better suited for a regex if the format remains the same. Doing a `split` over delimiters is fragile. This should do the trick: `(RHSA)\-\d{4}\:\d{4}`

Reply:
A regexp of Go is not fast, so I always avoid using it except for complicated patterns. Also, a regexp is often hard to understand. Therefore I used `split`, but I don't care about it. If you think we should use a regexp, I can replace this one with a regexp.

Reply:
Fair point. Let's keep it this way then.